src/coreclr/src/vm/amd64/asmhelpers.S

```diff
@@ -9,141 +9,180 @@
 #define real8 qword
 
-//
-//    file: profile.cpp
-//    typedef struct _PROFILE_PLATFORM_SPECIFIC_DATA
-//    {
-//        FunctionID *functionId; // function ID comes in the r11 register
-//        void       *rbp;
-//        void       *probersp;
-//        void       *ip;
-//        void       *profiledRsp;
-//        UINT64      rax;
-//        LPVOID      hiddenArg;
-//        UINT64      flt0;
-//        UINT64      flt1;
-//        UINT64      flt2;
-//        UINT64      flt3;
-//        UINT64      flt4;
-//        UINT64      flt5;
-//        UINT64      flt6;
-//        UINT64      flt7;
-//        UINT64      rdi;
-//        UINT64      rsi;
-//        UINT64      rdx;
-//        UINT64      rcx;
-//        UINT64      r8;
-//        UINT64      r9;
-//        UINT32      flags;
-//        UINT64      buffer[PROFILE_PLATFORM_SPECIFIC_DATA_BUFFER_SIZE];
-//    } PROFILE_PLATFORM_SPECIFIC_DATA, *PPROFILE_PLATFORM_SPECIFIC_DATA;
-
-// Buffer to copy structs sequentially
+#
+#    file: profile.cpp
+#    typedef struct _PROFILE_PLATFORM_SPECIFIC_DATA
+#    {
+#        FunctionID *functionId; # function ID comes in the r11 register
+#        void       *rbp;
+#        void       *probersp;
+#        void       *ip;
+#        void       *profiledRsp;
+#        UINT64      rax;
+#        LPVOID      hiddenArg;
+#        UINT64      flt0;
+#        UINT64      flt1;
+#        UINT64      flt2;
+#        UINT64      flt3;
+#        UINT64      flt4;
+#        UINT64      flt5;
+#        UINT64      flt6;
+#        UINT64      flt7;
+#        UINT64      rdi;
+#        UINT64      rsi;
+#        UINT64      rdx;
+#        UINT64      rcx;
+#        UINT64      r8;
+#        UINT64      r9;
+#        UINT32      flags;
+#        UINT64      buffer[PROFILE_PLATFORM_SPECIFIC_DATA_BUFFER_SIZE];
+#    } PROFILE_PLATFORM_SPECIFIC_DATA, *PPROFILE_PLATFORM_SPECIFIC_DATA;
+
+# On linux arguments can be passed in non-sequential registers. Integer arguments are
+# passed in sequential integer registers and floating point registers are passed in
+# sequential floating point registers. This presents a problem when we go to pass the
+# struct argument as a COR_PRF_FUNCTION_ARGUMENT_RANGE which expects the arguments are
+# in one contiguous range. This space is a scratch space that the ArgIterator can use
+# to copy the structs to so they are sequential.
 .equ SIZEOF_PROFILE_STRUCT_BUFFER, 0x8*16
-.equ SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA, 0x8*22 + 0x8 + SIZEOF_PROFILE_STRUCT_BUFFER // includes fudge to make FP_SPILL right
+
+# Space for register spilling
+.equ SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA, 0x8*22 + SIZEOF_PROFILE_STRUCT_BUFFER
+
+# Pad the frame size by 0x8 so when the xmm0 and xmm1 register store/restore happens
+# we can align to 16 and be guaranteed to not exceed the frame size
+.equ STACK_FUDGE_FACTOR, 0x8
+
+# Space to keep xmm0 and xmm1
+.equ SIZEOF_FP_ARG_SPILL, 0x10*2
+
+.equ OFFSETOF_FP_ARG_SPILL, SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA
```

Comment:
This needs to include STACK_FUDGE_FACTOR

PROFILE_PLATFORM_SPECIFIC_DATA is stored from [rsp, rsp+SIZEOF_PROFILE_SPECIFIC_DATA]. Xmm0 is stored at (rsp+OFFSETOF_FP_ARG_SPILL) & -16. Right now both of those constants are 08x*22 = 0xb0. If rsp was 8 we'd be storing Xmm0 starting at (0x8 + 0xb0) & -16 = 0xb0 but the last bytes of PROFILE_PLATFORM_SPECIFIC_DATA would be stored up until 0x8 + 0xb0 = 0xb8. The last 8 bytes of PROFILE_PLATFORM_SPECIFIC_DATA get overwritten when alignment slides xmm0 backwards 8 bytes.